sklearn/tree/_utils.pyx

```diff
@@ -301,3 +303,328 @@ cdef class PriorityHeap:
+        self.get_median(&current_median)
+        self.total_weight += weight
+
+        if data < current_median:
+            # inserting below the median, so increment k and
+            # then update self.sum_w_0_k accordingly by adding
+            # the weight that was added.
+            self.k += 1
+            # update sum_w_0_k by adding the weight added
+            self.sum_w_0_k += weight
+
+            # minimize k such that sum(W[0:k]) >= total_weight / 2
+            # minimum value of k is 1
+            while(self.k > 1 and (self.sum_w_0_k -
+                                  self.samples.get_weight_from_index(self.k-1)
+                                  >= self.total_weight / 2.0)):
+                self.k -= 1
+                self.sum_w_0_k -= self.samples.get_weight_from_index(self.k)
+
+            return 0
+
+        if data >= current_median:
+            # inserting above or at the median
+            # minimize k such that sum(W[0:k]) >= total_weight / 2
+            while(self.k < self.samples.size() and
+                  (self.sum_w_0_k < self.total_weight / 2.0)):
+                self.k += 1
+                self.sum_w_0_k += self.samples.get_weight_from_index(self.k-1)
+            return 0
+
+    cdef int remove(self, DOUBLE_t data, DOUBLE_t weight) nogil:
+        """Remove a value from the MedianHeap, removing it
+        from consideration in the median calculation
+        """
+        cdef double current_unweighted_median
+        cdef int return_value
+
+        return_value = self.samples.remove(data, weight)
+        self.update_median_parameters_post_remove(data, weight)
+        return return_value
+
+    cdef int pop(self, DOUBLE_t* data, DOUBLE_t* weight) nogil:
+        """Pop a value from the MedianHeap, starting from the
+        left and moving to the right.
+        """
+        cdef int return_value
+
+        # no elements to pop
+        if self.samples.size() == 0:
+            return -1
+
+        return_value = self.samples.pop(data, weight)
+        self.update_median_parameters_post_remove(data[0],
+                                                  weight[0])
+        return return_value
+
+    cdef int update_median_parameters_post_remove(self, DOUBLE_t data,
+                                                  DOUBLE_t weight) nogil:
+        """Update the parameters used in the median calculation,
+        namely `k` and `sum_w_0_k` after a removal"""
+        cdef DOUBLE_t current_median
+        # reset parameters because empty
+        if self.samples.size() == 0:
+            self.k = 0
+            self.total_weight = 0
+            self.sum_w_0_k = 0
+            return 0
+
+        # trivial case of one element.
+        if self.samples.size() == 1:
+            self.k = 1
+            self.total_weight -= weight
+            self.sum_w_0_k = self.total_weight
+            return 0
+
+        # get the current weighted median
+        self.get_median(&current_median)
```

Comment:
My personal preference is that if you're returning a single value, it shouldn't be done by writing to a pointer like this, but just return that single value.